OpenRA.Mods.Common/Traits/CashTrickler.cs

```diff
@@ -15,52 +15,73 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Lets the actor generate cash in a set periodic time.")]
-	class CashTricklerInfo : ITraitInfo
+	class CashTricklerInfo : UpgradableTraitInfo
 	{
 		[Desc("Number of ticks to wait between giving money.")]
-		public readonly int Period = 50;
+		public readonly int Interval = 50;
+
 		[Desc("Amount of money to give each time.")]
 		public readonly int Amount = 15;
-		[Desc("Whether to show the cash tick indicators (+$15 rising from actor).")]
+
+		[Desc("Whether to show the cash tick indicators rising from the actor.")]
 		public readonly bool ShowTicks = true;
-		[Desc("Amount of money awarded for capturing the actor.")]
-		public readonly int CaptureAmount = 0;
 
-		public object Create(ActorInitializer init) { return new CashTrickler(this); }
+		[Desc("How long to show the cash tick indicator when enabled.")]
+		public readonly int DisplayDuration = 30;
+
+		public override object Create(ActorInitializer init) { return new CashTrickler(this); }
 	}
 
-	class CashTrickler : ITick, ISync, INotifyCapture
+	class CashTrickler : UpgradableTrait<CashTricklerInfo>, ITick, ISync
 	{
 		readonly CashTricklerInfo info;
 		[Sync] int ticks;
+
 		public CashTrickler(CashTricklerInfo info)
+			: base(info)
 		{
 			this.info = info;
 		}
 
-		public void Tick(Actor self)
+		void ITick.Tick(Actor self)
 		{
+			if (IsTraitDisabled)
+				return;
+
 			if (--ticks < 0)
 			{
-				ticks = info.Period;
-				self.Owner.PlayerActor.Trait<PlayerResources>().GiveCash(info.Amount);
-				MaybeAddCashTick(self, info.Amount);
+				ticks = info.Interval;
+
+				if (info.Amount < 0)
+					RemoveCash(self, self.Owner, -info.Amount);
+				else
+					AddCash(self, self.Owner, info.Amount);
 			}
 		}
 
-		public void OnCapture(Actor self, Actor captor, Player oldOwner, Player newOwner)
+		void AddCashTick(Actor self, int amount)
 		{
-			if (info.CaptureAmount > 0)
-			{
-				newOwner.PlayerActor.Trait<PlayerResources>().GiveCash(info.CaptureAmount);
-				MaybeAddCashTick(self, info.CaptureAmount);
-			}
+			self.World.AddFrameEndTask(w => w.Add(
+				new FloatingText(self.CenterPosition, self.Owner.Color.RGB, FloatingText.FormatCashTick(amount), info.DisplayDuration)));
+		}
+
+		void AddCash(Actor self, Player owner, int amount)
+		{
+			owner.PlayerActor.Trait<PlayerResources>().GiveCash(amount);
+			if (info.ShowTicks)
+				AddCashTick(self, amount);
 		}
 
-		void MaybeAddCashTick(Actor self, int amount)
+		void RemoveCash(Actor self, Player owner, int amount)
 		{
+			var resources = owner.PlayerActor.Trait<PlayerResources>();
+			var totalCash = resources.Cash + resources.Resources;
+
+			// Check whether the amount of cash to be removed would exceed available player cash, in that case only remove all the player cash
+			var newAmount = totalCash - amount < 0 ? totalCash : amount;
```

Comment:
This can be condensed down to one line with `var newAmount = Math.Min(resources.Cash + resources.Resources, amount);`